### PR TITLE
e2e: add test code for queue controller

### PR DESCRIPTION
**Pull request by alcorj-mizar**
this add queue command close/open e2e scenario and queue job status transaction scenario

WIP #857

### CODE REVIEW

**Comment by alcorj-mizar** [comment]
> u are misunderstanding, the supported action for command are listed

Hi, @hzxuzhonghu , this pull request add multiple e2e cases, one of them is the command e2e case which only test close and open queue, that means command handler works well. I think other state will test by UT.
Another several e2e cases is to test pod group status transaction in queue, from inqueue to running, from running to pending, from running to unknown.

File: test/e2e/admission.go [hunk 1]
```diff
@@ -243,7 +243,9 @@ var _ = ginkgo.Describe("Job E2E Test: Test Admission service", func() {
 			},
 		}
 
-		_, err := ctx.vcclient.SchedulingV1beta1().PodGroups(ctx.namespace).Create(context.TODO(), pg, v1.CreateOptions{})
+		podGroup, err := ctx.vcclient.SchedulingV1beta1().PodGroups(ctx.namespace).Create(context.TODO(), pg, v1.CreateOptions{})
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		err = waitPodGroupPhase(ctx, podGroup, "Running")
```
Comment (hzxuzhonghu) [inline test/e2e/admission.go:248]: nit: don't hardcode it

File: pkg/cli/util/util.go [hunk 2]
```diff
@@ -93,6 +93,33 @@ func PopulateResourceListV1(spec string) (v1.ResourceList, error) {
 	return result, nil
 }
 
+// CreateQueueCommand executes a command such as open/close
+func CreateQueueCommand(queueClient *versioned.Clientset, ns, name string, action vcbus.Action) error {
```
Comment (hzxuzhonghu) [inline pkg/cli/util/util.go:97]: s/queueClient/vcClient

File: test/e2e/queue.go [hunk 3]
```diff
@@ -19,14 +19,184 @@ package e2e
...
 var _ = Describe("Queue E2E Test", func() {
+	It("Queue Command Close And Open With State Check", func() {
+		q1 := "queue-command-close"
+		ctx := initTestContext(options{
+			queues: []string{q1},
+		})
+		defer cleanupTestContext(ctx)
+
+		By("Close queue command check")
+		err := util.CreateQueueCommand(ctx.vcclient, "default", q1, "CloseQueue")
```
Comment (hzxuzhonghu) [inline test/e2e/queue.go:40]: s/ CloseQueue/ CloseQueueAction

File: test/e2e/queue.go [hunk 4]
```diff
@@ -19,14 +19,184 @@ package e2e
...
+		By("Open queue command check")
+		err = util.CreateQueueCommand(ctx.vcclient, "default", q1, "OpenQueue")
+		if err != nil {
+			Expect(err).NotTo(HaveOccurred(), "Error send open queue command")
+		}
+
+		err = waitQueueStatus(func() (bool, error) {
+			queue, err := ctx.vcclient.SchedulingV1beta1().Queues().Get(context.TODO(), q1, metav1.GetOptions{})
+			Expect(err).NotTo(HaveOccurred(), "Get queue %s failed", q1)
+			return queue.Status.State == "Open", nil
```
Comment (hzxuzhonghu) [inline test/e2e/queue.go:61]: prevent hardcoding

File: test/e2e/queue.go [hunk 5]
```diff
@@ -19,14 +19,184 @@ package e2e
...
+
+			q1 = "queue-jobs-status-transaction"
+			ctx = initTestContext(options{
+				queues: []string{q1},
+			})
+			podNamespace = ctx.namespace
+			slot := oneCPU
+			rep = clusterSize(ctx, slot)
+
+			if rep < 4 {
```
Comment (hzxuzhonghu) [inline test/e2e/queue.go:84]: This not friendly, should not assume where to run this case.  We can skip this case if the resource can not satisfy

File: test/e2e/queue.go [hunk 6]
```diff
@@ -19,14 +19,184 @@ package e2e
...
+				err = waitQueueStatus(func() (bool, error) {
+					queue, err := ctx.vcclient.SchedulingV1beta1().Queues().Get(context.TODO(), q1, metav1.GetOptions{})
+					Expect(err).NotTo(HaveOccurred(), "Get queue %s failed", q1)
+					return queue.Status.Running > 0, nil
+				})
+				Expect(err).NotTo(HaveOccurred(), "Error wait for queue running")
+			})
+		})
+
+		AfterEach(func() {
```
Comment (hzxuzhonghu) [inline test/e2e/queue.go:129]: Move this after `BeforeEach`?

File: test/e2e/queue.go [hunk 7]
```diff
@@ -19,14 +19,184 @@ package e2e
...
+					queue, err := ctx.vcclient.SchedulingV1beta1().Queues().Get(context.TODO(), q1, metav1.GetOptions{})
+					Expect(err).NotTo(HaveOccurred(), "Get queue %s failed", q1)
+					return queue.Status.Pending > 0, nil
+				})
+				Expect(err).NotTo(HaveOccurred(), "Error wait for queue Pending")
+			})
+		})
+
+		Context("When some of pod deleted and lackof resources pod group status should transaction from Running to Unknown", func() {
+			It("Transate from running to unknown should succeed", func() {
```
Comment (hzxuzhonghu) [inline test/e2e/queue.go:162]: Transate /transform?

File: test/e2e/queue.go [hunk 8]
```diff
@@ -19,14 +19,184 @@ package e2e
...
+				By("Verify queue have pod groups running")
+				err := waitQueueStatus(func() (bool, error) {
+					queue, err := ctx.vcclient.SchedulingV1beta1().Queues().Get(context.TODO(), q1, metav1.GetOptions{})
+					Expect(err).NotTo(HaveOccurred(), "Get queue %s failed", q1)
+					return queue.Status.Running > 0, nil
+				})
+				Expect(err).NotTo(HaveOccurred(), "Error wait for queue running")
+
+				By("Delete some of pod which will case pod")
```
Comment (hzxuzhonghu) [inline test/e2e/queue.go:171]: what does this mean

Comment (alcorj-mizar) [reply test/e2e/queue.go:171]: Forgot to complete it, i will fix that.

File: test/e2e/queue.go (hunk 5 above)
Comment (alcorj-mizar) [reply test/e2e/queue.go:84]: Yeah, but i think when resource is not satisfy, we should remind the tester, then he(she) can skip it.
Slot is too big here, i was intended to make the slot half of cpu, and then many laptop can support the case.

File: test/e2e/queue.go (hunk 6 above)
Comment (alcorj-mizar) [reply test/e2e/queue.go:129]: OK

File: test/e2e/queue.go (hunk 7 above)
Comment (alcorj-mizar) [reply test/e2e/queue.go:162]: OK

File: test/e2e/queue.go (hunk 4 above)
Comment (alcorj-mizar) [reply test/e2e/queue.go:61]: OK, it's better.

File: test/e2e/admission.go (hunk 1 above)
Comment (alcorj-mizar) [reply test/e2e/admission.go:248]: OK

File: test/e2e/queue.go (hunk 5 above)
Comment (alcorj-mizar) [reply test/e2e/queue.go:87]: The Errorf was rewritten to more friendly :)

File: test/e2e/queue.go [hunk 9]
```diff
@@ -19,14 +19,187 @@ package e2e
...
+		err = waitQueueStatus(func() (bool, error) {
+			queue, err := ctx.vcclient.SchedulingV1beta1().Queues().Get(context.TODO(), q1, metav1.GetOptions{})
+			Expect(err).NotTo(HaveOccurred(), "Get queue %s failed", q1)
+			return queue.Status.State == schedulingv1beta1.QueueStateOpen, nil
+		})
+		Expect(err).NotTo(HaveOccurred(), "Wait for reopen queue %s failed", q1)
+
+	})
+
+	Describe("Queue Job Status Transaction", func() {
```
Comment (hzxuzhonghu) [inline test/e2e/queue.go:70]: What do u mean by "Queue Job Status Transaction"

File: test/e2e/queue.go [hunk 10]
```diff
@@ -19,14 +19,187 @@ package e2e
...
+		var ctx *testContext
+		var q1 string
+		var podNamespace string
+		var rep int32
+		var firstJobName string
+		By("Prepare 3 job")
```
Comment (hzxuzhonghu) [inline test/e2e/queue.go:76]: Can you explain why prepare 3 jobs here?

File: test/e2e/queue.go (hunk 9 above)
Comment (hzxuzhonghu) [reply test/e2e/queue.go:70]: transition?

File: test/e2e/queue.go [hunk 11]
```diff
@@ -19,14 +19,187 @@ package e2e
...
+				spec.queue = q1
+				createJob(ctx, spec)
+			}
+		})
+
+		AfterEach(func() {
+			cleanupTestContext(ctx)
+		})
+
+		Context("When transform from Inqueue to Running", func() {
```
Comment (hzxuzhonghu) [inline test/e2e/queue.go:116]: I donot think Context is necessary here, right?

File: test/e2e/queue.go [hunk 12]
```diff
@@ -19,14 +19,187 @@ package e2e
...
+				Expect(err).NotTo(HaveOccurred(), "Failed waiting for pods")
+
+				clusterPods, err := ctx.kubeclient.CoreV1().Pods(podNamespace).List(context.TODO(), metav1.ListOptions{})
+				for _, pod := range clusterPods.Items {
+					if pod.Status.Phase == corev1.PodRunning {
+						err = ctx.kubeclient.CoreV1().Pods(podNamespace).Delete(context.TODO(), pod.Name, metav1.DeleteOptions{})
+						Expect(err).NotTo(HaveOccurred(), "Failed to delete pod %s", pod.Name)
+						podDeleteNum = podDeleteNum + 1
+					}
+					if podDeleteNum >= int(rep/2) {
```
Comment (hzxuzhonghu) [inline test/e2e/queue.go:187]: I can hardly understand what these above codes are doing

File: test/e2e/queue.go (hunk 9 above)
Comment (alcorj-mizar) [reply test/e2e/queue.go:70]: Transition may more precision, will update it.

File: test/e2e/queue.go (hunk 11 above)
Comment (alcorj-mizar) [reply test/e2e/queue.go:116]: Not necessary, but better use it in one behavior different circumstances.

File: test/e2e/queue.go (hunk 12 above)
Comment (alcorj-mizar) [reply test/e2e/queue.go:187]: This cases is not easy to reappear， i need to make sure there is some of pod is in running state, and locate then and delete (restart) them which will case an unknown state.

File: test/e2e/queue.go (hunk 10 above)
Comment (alcorj-mizar) [reply test/e2e/queue.go:76]: To make sure the resources has been fully loaded by pods.

**Pull request merged**